keps/sig-apps/4443-configurable-pod-failure-policy-reasons/README.md

```diff
@@ -0,0 +1,971 @@
+so the simulation should be restarted.
+
+When a Job fails due to a pod failing with exit code 2, I want my job management software to leave the Job in
+a failed state.
+
+When a Job fails due to a pod failing with exit code 3, I want my job management software to to restart the Job.
+
+**Example JobSet for this use case**:
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: restart-job-example
+  annotations:
+    alpha.jobset.sigs.k8s.io/exclusive-topology: {{topologyDomain}} # 1:1 job replica to topology domain assignment
+spec:
+  failurePolicy:
+    rules:
+    # If Job fails due to a pod failing with exit code 2, leave it in a failed state.
+    - action: FailJob
+      targetReplicatedJobs:
+      - simulations
+      onJobFailureReasons:
+      - PodFailurePolicy_ExitCode2  # Job failure reason format: PodFailurePolicy_{ruleName}
+    # If Job fails due to a pod failing with exit code 3, restart that Job.
+    - action: RestartJob
+      targetReplicatedJobs:
+      - simulations
+      onJobFailureReasons:
+      - PodFailurePolicy_ExitCode3  # Job failure reason format: PodFailurePolicy_{ruleName}
+    maxRestarts: 10
+  replicatedJobs:
+  - name: simulations
+    replicas: 10
+    template:
+      spec:
+        parallelism: 1
+        completions: 1
+        backoffLimit: 0
+        # Pod failure policy rules, the names of which are referenced in the JobSet failure policy.
+        podFailurePolicy:
+          rules:
+          - name: ExitCode2
+            action: FailJob
+            onExitCodes:
+              containerName: main
+              operator: In
+              values: [2]
+          - name: ExitCode3
+            action: FailJob
+            onExitCodes:
+              containerName: main
+              operator: In
+              values: [3]
+        template:
+          spec:
+            restartPolicy: Never
+            containers:
+            - name: main
+              image: python:3.10
+              command: ["..."]
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+There is a risk to making a field that was previously exclusively managed by the controller,
+to now being configurable by the user. However, as described in the validation section below,
+we are validating against malformed/invalid inputs.
+
```

Comment:
Aren't pod failure policy rules evaluated in order, so if 2 rules match a given pod failure then only the one which appears first in the list of rules will be triggered?

Reply:
True, but it might still be confusing for users who do not read that part. It might be good enough to add it to the `Notes/Constraints/Caveats` section. wdyt?

Reply:
Added a note on this in the `Notes/Constraints/Caveats` section.